mmdeploy/codebase/mmdet3d/deploy/voxel_detection.py

```diff
@@ -50,107 +67,152 @@ def build_pytorch_model(self,
+        data.append(data_)
+
+        collate_data = pseudo_collate(data)
+        data[0]['inputs']['points'] = data[0]['inputs']['points'].to(
+            self.device)
+
+        if preproc is not None:
+            collate_data = preproc(collate_data, False)
+            del collate_data['inputs']['voxels']['voxel_centers']
+
+        voxels = collate_data['inputs']['voxels']
+        inputs = [voxels['voxels'], voxels['num_points'], voxels['coors']]
+        return collate_data, inputs
 
     def visualize(self,
+                  image: Union[str, np.ndarray],
                   model: torch.nn.Module,
-                  image: str,
                   result: list,
                   output_file: str,
-                  window_name: str,
+                  window_name: str = '',
                   show_result: bool = False,
-                  score_thr: float = 0.3):
-        """Visualize predictions of a model.
+                  draw_gt: bool = False,
+                  data_samples=None,
+                  **kwargs):
+        """_summary_
 
         Args:
-            model (nn.Module): Input model.
-            image (str): Pcd file to draw predictions on.
-            result (list): A list of predictions.
-            output_file (str): Output file to save result.
-            window_name (str): The name of visualization window. Defaults to
-                an empty string.
-            show_result (bool): Whether to show result in windows, defaults
-                to `False`.
-            score_thr (float): The score threshold to display the bbox.
-                Defaults to 0.3.
+            image (Union[str, np.ndarray]): _description_
+            result (list): _description_
+            output_file (str): _description_
+            window_name (str, optional): _description_. Defaults to ''.
+            show_result (bool, optional): _description_. Defaults to False.
+            draw_gt (bool, optional): _description_. Defaults to False.
         """
-        from mmdet3d.apis import show_result_meshlab
-        data = VoxelDetection.read_pcd_file(image, self.model_cfg, self.device)
-        show_result_meshlab(
-            data,
-            result,
-            output_file,
-            score_thr,
+        cfg = self.model_cfg
+        visualizer = super().get_visualizer(window_name, output_file)
+        visualizer.dataset_meta = _get_dataset_metainfo(cfg)
+
+        # show the results
+        _, data_input = self.create_input(pcd=image)
+
+        if get_backend(self.deploy_cfg) != Backend.PYTORCH:
+            predictions = VoxelDetectionModel.postprocess(
+                model_cfg=self.model_cfg,
+                deploy_cfg=self.deploy_cfg,
+                outs=result,
+                metas=data_samples)
+            prediction = predictions[0]
+        else:
+            prediction = result
+
+        visualizer.add_datasample(
+            window_name,
+            dict(points=data_input['points'][0]),
```

Comment:
data_input is a list.

Reply:
It is worth to mention that in `aps/visualize.py`, visualize has been wrapped inside the try catch, so whatever visualize error would be treated as headless device error.

Reply:
done.